File: workspace-raspi/net/src/http_pi/contextProcessor.java
/**
 * Raspberry Pi java http server
 * v1.8 update 2015.2.28
 * contextProcessor.java
 *  コンテンツへのリクエストを処理します。
 */
package http_pi;

import java.io.BufferedInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.Date;

/**
 * @author nishi
 *
 */
public class contextProcessor {
	private http_comLib _http_comLib;
	private boolean _ka;
	private byte[] buff;
	/**
	 * contextProcessor(boolean ka,byte[] _buff,http_comLib m_http_comLib)
	 */
	public contextProcessor(boolean ka,byte[] _buff,http_comLib m_http_comLib){
		_ka=ka;
		buff=_buff;
		_http_comLib=m_http_comLib;
	}
	/*
	 * void set_ka(boolean ka)
	 */
	public void set_ka(boolean ka){
		_ka=ka;
		_http_comLib.set_ka(ka);
	}
	public void exec(Socket sock,HttpReader hrd,String indexname){
		File file;
		boolean flg=false;
		boolean img_f=false;
		boolean expir_f=false;

		String contentType;	//コンテントタイプ
		String	version = hrd.version;

		System.out.println("contentProcessor::exec :#1 indexname="+indexname);
		try {
			//OutputStream outs = new BufferedOutputStream(sock.getOutputStream());
			//BufferedOutputStream outs = new BufferedOutputStream(sock.getOutputStream());
			DataOutputStream outs = new DataOutputStream(sock.getOutputStream());
			//Writer out = new OutputStreamWriter(outs);
			
			if(indexname.equals("")){
				indexname="index.html";
				file = new File(Defs.doc_rootDir +"/"+ indexname);
				flg = file.canRead();
				if(flg != true){
					indexname="index.htm";
				}
			}
			contentType = "text/html";
			String adr = Defs.doc_rootDir +"/"+ indexname;
			System.out.println("contentProcessor::exec :#2 adr="+adr);

			//System.out.println("requestProcessor :#5-1 socket.getKeepAlive()="+sock.getKeepAlive());
			//System.out.println("requestProcessor :#5-1 socket.getSendBufferSize()="+sock.getSendBufferSize());
			if(adr.endsWith(".gif")){
				contentType = "image/gif";
				img_f=true;
				expir_f=true;
			}
			else if(adr.endsWith(".jpg")){
				contentType = "image/jpg";
				img_f=true;
				expir_f=true;
			}
			else if(adr.endsWith(".png")){
				contentType = "image/png";
				img_f=true;
				expir_f=true;
			}
			else if(adr.endsWith(".css")){
				contentType = "text/css";
				expir_f=true;
			}
			else if(adr.endsWith(".js")){
				contentType = "text/javascript";
				expir_f=true;
			}
			file = new File(adr);
			flg = file.canRead();
			Date now;

			if(flg){
				System.out.println("contentProcessor::exec :#3 output to client ad="+adr);
				System.out.println("contentProcessor::exec :#4 output file size="+file.length());
				
				if(version.startsWith("HTTP/")){	//MIME header read
					System.out.println("contentProcessor::exec :#5 output http response with contentType="+contentType);
					String hd_s=_http_comLib.mkHttpHeaderOK_file(hrd.method,img_f,expir_f,file,contentType);
					outs.writeBytes(hd_s);
					outs.flush();
				}
				// GET Method
				if(hrd.method.equals("GET")==true){
					put_file_bin(file,outs);
					outs.flush();
				}
			}
			else{	//ファイル未検出
				System.out.println("contentProcessor::exec :#7 file nothing");
				now = new Date();
				String hd_s="";
				if(version.startsWith("HTTP/")){	//MIME header read
					hd_s="HTTP/1.1 404 File Not Found\r\n"
						+"Date: "+now+"\r\n"
						+"Server: HTTP 1.1\r\n"
						+"Content-type: text/html\r\n\r\n";	// 注）最後は \r\nが２個必要
					outs.writeBytes(hd_s);
					outs.flush();
				}
				hd_s="<HTML>\r\n"
					+"<HEAD><TITLE>File Not Found</TITLE></HEAD>\r\n"
					+"<BODY>\r\n"
					+"<H1>HTTP Error 404: File Not Found</H1>\r\n"
					+"<BODY>\r\n";
				outs.writeBytes(hd_s);
				outs.flush();
			}
			//outs.close();
		}
		catch (IOException e) {
			System.out.println("contentProcessor::exec :#90 e="+e);
		}
	}
	/*
	 * void put_file_bin(String adr,DataOutputStream outs)
	 */
	private void put_file_bin(File file,DataOutputStream outs){
		try{
			BufferedInputStream in = new BufferedInputStream(
		            new FileInputStream(file));
		    int len;
		    while ((len = in.read(buff)) != -1) {
				//System.out.println("put_file_bin :#1 output len="+len);
		       outs.write(buff, 0, len);
		    }
		    in.close();
		}
		catch(IOException e){
		}
	}
}

File: workspace-raspi/http_pic_t.v1.3.2.1/src/contextProcessor.c
/*
 * contextProcessor.c
 * Version 1.1
 * Update on: 2014/03/26
 *      Author: nishi
 */
#include <pthread.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
//#include <sys/socket.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
//#include <netinet/in.h>
//#include <signal.h>
#include <stdbool.h>

#include "constDef.h"
#include "tableDef.h"
#include "serverConfig.h"
#include "httpReader.h"
#include "httpWriter.h"
#include "com_lib.h"

void contextProcessor_nothing(HttpReq *httpReq){

	char hd_s[]="HTTP/1.1 404 File Not Found\r\n";
		//+"Date: "+now+"\r\n"
	char hd_s2[]="Server: HTTP 1.1\r\n";
	char hd_s3[]="Content-type: text/html\r\n\r\n";	// 注）最後は \r\nが２個必要

	write(httpReq->sockfd,hd_s,strlen(hd_s));
	write(httpReq->sockfd,hd_s2,strlen(hd_s3));
	write(httpReq->sockfd,hd_s3,strlen(hd_s3));

	char hd_s4[]="<HTML>\r\n";
	char hd_s5[]="<HEAD><TITLE>File Not Found</TITLE></HEAD>\r\n";
	char hd_s6[]="<BODY>\r\n";
	char hd_s7[]="<H1>HTTP Error 404: File Not Found</H1>\r\n";
	char hd_s8[]="<BODY>\r\n";
	write(httpReq->sockfd,hd_s4,strlen(hd_s4));
	write(httpReq->sockfd,hd_s5,strlen(hd_s5));
	write(httpReq->sockfd,hd_s6,strlen(hd_s6));
	write(httpReq->sockfd,hd_s7,strlen(hd_s7));
	write(httpReq->sockfd,hd_s8,strlen(hd_s8));
	//close(httpReq->sockfd);
}
/*
 * void put_file_bin(HttpReq *httpReq,char *file_ph,int o_fd)
 * 実際のコンテンツの出力
 */
void put_file_bin(HttpReq *httpReq,char *file_ph,int o_fd){
	int i_fd;
	int r_lng;
	i_fd=open(file_ph,O_RDONLY);
	//printf("contextProcessor::put_file_bin() : #1 i_fd=%d\n",i_fd);
	if(i_fd != -1){
		while(1){
			r_lng=read(i_fd,httpReq->buff,sizeof(httpReq->buff));
			if(r_lng>0){
				if(write(o_fd,httpReq->buff,r_lng) != r_lng){
					printf("contextProcessor::put_file_bin():#3 o_fd=%d write error\n",o_fd);
					break;
				}
				//最後 +1 の読み込みをなくしてみます。2014.3.30
				if(r_lng < sizeof(httpReq->buff)){
					break;
				}
			}
			else if(r_lng==0){
				break;
			}
			else{
				printf("contextProcessor::put_file_bin() : #4 read error\n");
				break;
			}
		}
		close(i_fd);
	}
	else{
		printf("contextProcessor::put_file_bin():#10 open error\n");
	}
}
/*
 * contextProcessor(HttpReq *httpReq,char *index_name)
 * コンテキストプロセッサー
 * html,imgの処理を行います
 */
void contextProcessor(HttpReq *httpReq,char *index_name){
	StrBuff *strBuff;
	char *contentType;	// const use
	char *version;	// const use

	char *m_indexname;	// rq_malloc use
	char *m_file_s;		// rq_malloc use
	char *m_file_s2;	// rq_malloc use
	char *m_dir;		// rq_malloc use

	struct stat  st;
	bool img_f;
	char *hd_s;

	char f_size_s[20];
	int rc;

	strBuff=&(httpReq->sbuff);

	//printf(">contextProcessor():#1 called %d,%d\n",httpReq->seq_i,httpReq->sockfd);

	//printf("contextProcessor():#1 index_name=%s\n",index_name);

	m_dir=rq_strcat(strBuff,defs.doc_rootDir,"/");
	m_file_s=rq_strcat(strBuff,m_dir,index_name);
	//ファイル無し
	if (stat(m_file_s, &st) != 0){
		m_indexname=rq_strcpy(strBuff,index_name);
		//printf("contextProcessor():#1 passed!\n");
	}
	else{
		//printf("contextProcessor():#2 m_file_s=%s\n",m_file_s);
		// ディレクトリーです
		if(S_ISDIR(st.st_mode) == 1){
			// 最後が "/" です
			if(str_endsWith(index_name, "/") == 0){
				//printf("contextProcessor():#3 passed!\n");
				strcpy(httpReq->buff,m_file_s);
				strcat(httpReq->buff,"index.html");
				//ファイル無し
				if (stat(httpReq->buff, &st) != 0){
					m_indexname=rq_strcat(strBuff,index_name,"index.htm");
				}
				else{
					m_indexname=rq_strcat(strBuff,index_name,"index.html");
				}
			}
			else{
				//printf("contextProcessor():#4 passed!\n");
				strcpy(httpReq->buff,m_file_s);
				strcat(httpReq->buff,"/index.html");
				//ファイル無し
				if (stat(httpReq->buff, &st) != 0){
					m_indexname=rq_strcat(strBuff,index_name,"/index.htm");
				}
				else{
					m_indexname=rq_strcat(strBuff,index_name,"/index.html");
				}
			}
		}
		else{
			//printf("contextProcessor():#5 passed!\n");
			m_indexname=rq_strcpy(strBuff,index_name);
		}
	}

	contentType = "text/html";
	version=httpReq->version;

	m_file_s2=rq_strcat(strBuff,m_dir,m_indexname);
#ifndef _SILENT
	//printf("contextProcessor() :#4 %d,file_s=%s\n",httpReq->seq_i,m_file_s2);
	if(str_endsWith(m_indexname,".html") == 0 || str_endsWith(m_indexname,".htm") == 0 ||
			str_endsWith(m_indexname,".css") == 0){
		printf("contextProcessor() :#4 %d,%d,file_s=%s\n",httpReq->seq_i,httpReq->sockfd,index_name);
	}
	else{
		printf("contextProcessor() :#4 %d,%d ..\n",httpReq->seq_i,httpReq->sockfd);
	}
#endif
	//ファイル無し
	if (stat(m_file_s2, &st) != 0){
#ifndef _SILENT
		printf("contextProcessor() :#5 file nothing file_s=%s\n",m_file_s2);
#endif
		contextProcessor_nothing(httpReq);
	}
	//ファイル有
	else{
		if(str_endsWith(m_indexname,".gif") == 0){
			contentType = "image/gif";
			img_f=true;
		}
		else if(str_endsWith(m_indexname,".jpg")==0){
			contentType = "image/jpg";
			img_f=true;
		}
		else if(str_endsWith(m_indexname,".css")==0){
			contentType = "text/css";
		}
		else if(str_endsWith(m_indexname,".js")==0){
			contentType = "text/javascript";
		}
		else if(str_endsWith(m_indexname,".png")==0){
			contentType = "image/png";
			img_f=true;
		}
		rc=0;
		if(str_startsWith(version,"HTTP/")==0){	//MIME header read
			// PUT Http Header OK
			rc=httpWriterOK(httpReq,contentType,img_f,st.st_size,&st.st_mtime);
		}
		if(rc==0){
			//printf("contextProcessor() :#10  call put_file_bin %d,%d\n",httpReq->seq_i,httpReq->sockfd);
			put_file_bin(httpReq,m_file_s2,httpReq->sockfd);
		}
	}
#ifndef _SILENT
	printf("<contextProcessor() :#90 end %d,%d\n",httpReq->seq_i,httpReq->sockfd);
#endif
}

